lightkurve/mast.py

```diff
@@ -61,86 +69,242 @@ def search_kepler_products(target):
+            month = [month]
+
+    #Grab the products
+    products = search_kepler_products(target=target, filetype=filetype, cadence=cadence,
+                                      quarter=quarter, campaign=campaign,
+                                      radius=radius)
+    if len(products) < 1:
+        raise ArchiveError("No {} File found for {} at MAST.".format(filetype, target))
+
+
+    #Limit to the correct number of hits based on ID. If there are multiple versions
+    #of the same ID, this shouldn't count towards the limit.
+    if targetlimit is not None:
+        ids = np.asarray([p.split('/')[-1].split('-')[0].split('_')[0][4:]
+                          for p in products['dataURI']], dtype=int)
+        if len(np.unique(ids)) < targetlimit:
+            logging.warning('Target return limit set to {} '
+                            'but only {} unique targets found. '
+                            'Trying increasing search radius. '
+                            '(radius currently set to {} arcseconds)'
+                            ''.format(targetlimit, len(np.unique(ids)), radius))
+        okids = ids[np.sort(np.unique(ids, return_index=True)[1])[0:targetlimit]]
+        mask = np.zeros(len(ids), dtype=bool)
+        idsort = np.zeros(len(ids))
+        for idx, o in enumerate(okids):
+            p = ids == o
+            idsort[p] = int(idx)
+            mask |= p
+        products['ids'] = idsort
+        products = products[mask]
+
+    #For Kepler short cadence data there are additional rules, so find anywhere
+    #where there is short cadence data...
+    scmask = np.asarray(['Short' in d for d in products['description']]) &\
+             np.asarray(['kplr' in d for d in products['dataURI']])
+    if np.any(scmask):
+        #Error check the user if there's multiple months and they didn't ask
+        #for a specific one
+        if month is None:
+            raise ArchiveError("Found {} different Target Pixel Files "
+                               "for target {} in Quarter {}. "
+                               "Please specify the month (1, 2, or 3)."
+                               "".format(len(products), target, quarter))
+        #Get the short cadence date lookup table.
+        table = ascii.read('/Users/ch/K2/repos/lightkurve/lightkurve/data/short_cadence_month_lookup.csv')
+        #Grab the dates of each of the short cadence files and make sure they
+        #all use the correct month
+        finalmask = np.ones(len(products), dtype=bool)
+        for c in np.unique(products[scmask]['qoc']):
+            ok = (products['qoc'] == c) & (scmask)
+            mask = np.zeros(np.shape(products[ok])[0], dtype=bool)
+            for m in month:
+                udate = (table['Date'][np.where((table['Month'] == m) & (table['Quarter'] == c))[0][0]])
+                mask |= np.asarray(products['dates'][ok]) == udate
+            finalmask[ok] = mask
+        products = products[finalmask]
```

Comment:
I wonder if we can simplify this somehow.  This is going to be tricky to understand and maintain in the future.